### PR TITLE
Improve support for combo boxes in all browsers

### LINKED ISSUE #8664: Unable to close combo boxes in Google Chrome
### Steps to reproduce:
1. Open a page in Google Chrome containing a combo box (select of size 1).
2. In browse mode, arrow to the combo box.
3. Press alt+downArrow to expand the combo box
Note that the combo box opens and NVDA switches to focus mode.
4. Press either escape or alt+upArrow to close the combo box.

### Actual behavior:
NVDA switches to browse mode and the combo box does not close.

### Expected behavior:
NVDA should switch to browse mode and the combo box should close.

Technical info:
NVDA thinks the list and list items inside the combo box are a part of the browseMode treeInterceptor.

#### NVDA Installed/portable/running from source:
Installed

#### NVDA version:
2018.3beta3
#### Windows version:
10.0.17744

#### Name and version of other software in use when reproducing the issue:
Google Chrome 70.0.3530.0 (Official Build) canary (64-bit) (cohort: Clang-64)

#### Does the issue still occur after restarting your PC?
yes

#### Have you tried any other versions of NVDA?
2018.2.1

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #8664 

### Summary of the issue:
Currently in NVDA, the user experience when using combo boxes on the web significantly differs by browser, plus some of the original code written to handle combo boxes has now broken in multiple browsers.
For example:
*In Firefox, collapsing a combo box works, but browseMode is not re-enabled.
* In Chrome 68, focusing a collapsed combo box ends up focusing the active list item within the combo box, even though it is collapsed.
* In Chrome 68 and 70, collapsing a combo box with escape/enter/alt+upArrow simply switches to browseMode and does not collapse the combo box.
* In Edge, Expanding the combo box and then arrowing through the options reports nothing and an exception occurs.
 
### Description of how this pull request fixes the issue:
This PR standardizes the way we handle combo boxes in browseMode, so that:
* Anything inside a combo box is treated as not being in the browseMode document. This means that if the combo box is expanded and focus is inside, browseMode completely ignores it.
* Alt+downArrow when in browseMode expands the combo box and switches to focus mode, but also remembers the combo box, so that the next focus event in browseMode (if it is for the combo box), switches back to browseMode again (as this means the combo box has been collapsed). Technically there is no need to toggle focus mode and browseMode at all, but this keeps UX compatible with previous NVDA versions.

This PR also partly works around a bug in Chrome 68 and below where focusing a collapsed combo box focuses the active list item within the combo box. This has been fixed in Chrome 70.

This PR also fixes an issue in Edge where list items in the combo box were being treated were being treated as part of the document but with a broken textRange.
 
### Testing performed:
Tested a standard HTML select (size=1) in Firefox 52, Firefox Nightly, Chrome 68 and Chrome 70.
Arrowed to the combo box in browse mode. Pressed alt+downArrow. Arrowed through the list. And closed the combo box with either enter, escape or alt+upArrow.

### Known issues with pull request:
Combo boxes in Chrome 68 and below are still not perfect, but no worse than what they were before this PR. 
 
### Change log entry:
Bug fixes:
* Interacting with combo boxes on the web in Firefox, Chrome and Internet Explorer has been improved.

### CODE REVIEW

File: source/eventHandler.py [hunk 1]
```diff
@@ -143,6 +143,9 @@ def executeEvent(eventName,obj,**kwargs):
 	@param kwargs: Additional event parameters as keyword arguments.
 	"""
 	try:
+		# Allow NVDAObjects to redirect focus events to another object of their choosing.
+		if eventName=="gainFocus" and obj.focusRedirect:
+			obj=obj.focusRedirect
```
Comment (leonardder) [inline source/eventHandler.py]: focusRedirect is used in the powerpnt appModule. We might have to make sure that this does not break. Having said that, I really like this being handled on the events level!

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
I found the following issue in Firefox:

1. Explicitly disable browse mode with NVDA+space
2. Expand and collapse a combo box

Expected: Browse mode stays off
Actual: Browse mode enables after collapsing the combo box.

Also, I wonder whether it would make sense to have the activate script expand a combo box (for example when pressing space on a combo box when in browse mode).

**Comment by michaelDCurran** [comment]
@leonardder: I fixed the issue you found with  NVDA switching back to browseMode after collapsing a combo box in focus mode. 
In regards to your query on whether we should automatically expand a combobox when pressing enter/space in browseMode: I don't think we should at this point in time as it is certainly valid to press enter/space on a combo box and then use the arrow keys to change the combo box value with out expanding it. This would no longer be possible, unless you explicitly switched to focus mode.

**Comment by leonardder** [comment]
Your latter point makes sense indeed.

**Comment by michaelDCurran** [comment]
Is there anything else I have to address?

**Comment by leonardder** [comment]
Sorry it took me some time to get back to this.

I'm now observing the following in Firefox:

1. In firefox, select a combo box in browse mode.
2. Expand with alt+down arrow
3. Press tab

Expected: NVDA stays in focus mode and moves to the control in tab order after the combo box. This is how Edge and IE work.
Actual: The combo box is collapsed and focus stays on the combo box. Browse mode is reactivated. IN Chrome 68, the focus stays on the combo box as well, but Chrome stays in focus mode. Another tab brings you to the expected next item in the tab order.

Apart from that, the code looks ok.

**Comment by michaelDCurran** [comment]
It seems that this is a deliberate choice by Firefox and Chrome to 
simply collapse the combo box when tab is pressed.

I don't really believe we should work around this one.

**Review by LeonarddeR — DISMISSED** [review]
IN that case, I agree.

**Pull request merged**